Cabal/Distribution/Simple.hs

```diff
@@ -328,7 +334,10 @@ haddockAction hooks flags args = do
 cleanAction :: UserHooks -> CleanFlags -> Args -> IO ()
 cleanAction hooks flags args = do
     distPref <- findDistPrefOrDefault (cleanDistPref flags)
-    let flags' = flags { cleanDistPref = toFlag distPref }
+
+    lbi <- getBuildConfig hooks verbosity distPref
+    let flags' = flags { cleanDistPref = toFlag distPref
+                       , cleanCabalFilePath = maybeToFlag (cabalFilePath lbi)}
```

Comment:
This is essentially the offending part. We try to `getBuildConfig` in 338, and then try to use the local build info here. But this might just fail, if the build Config is too old. (In most cases we should be able to read it, as we produce it with the same cabal, that we try to clean it with, unless we bootstrap...).